pymc_experimental/model_transform.py

```diff
@@ -0,0 +1,175 @@
+from pytensor.graph.basic import get_var_by_name
+from pytensor.graph.rewriting.basic import in2out
+
+from pymc_experimental.utils.model_fgraph import (
+    ModelObservedRV,
+    ModelValuedVar,
+    fgraph_from_model,
+    model_free_rv,
+    model_from_fgraph,
+    model_named,
+)
+
+__all__ = (
+    "forecast_timeseries",
+    "uncensor",
+)
+
+
+@node_rewriter(tracks=[ModelValuedVar])
+def uncensor_node_rewrite(fgraph, node):
+    """Rewrite that replaces censored variables by uncensored ones"""
+
+    (
+        censored_rv,
+        value,
+        *dims,
+    ) = node.inputs
+    if not isinstance(censored_rv.owner.op, CensoredRV):
+        return
+
+    model_rv = node.outputs[0]
+    base_rv = censored_rv.owner.inputs[0]
+    uncensored_rv = node.op.make_node(base_rv, value, *dims).default_output()
+    uncensored_rv.name = f"{model_rv.name}_uncensored"
+    return [uncensored_rv]
+
+
+uncensor_rewrite = in2out(uncensor_node_rewrite)
+
+
+def uncensor(model: Model) -> Model:
+    """Replace censored variables in the model by uncensored equivalent.
+
+    Replaced variables have the same name as original ones with an additional "_uncensored" suffix.
+
+    .. code-block:: python
+
+        import arviz as az
+        import pymc as pm
+        from pymc_experimental.model_transform import uncensor
+
+        with pm.Model() as model:
+            x = pm.Normal("x")
+            dist_raw = pm.Normal.dist(x)
+            y = pm.Censored("y", dist=dist_raw, lower=-1, upper=1, observed=[-1, 0.5, 1, 1, 1])
+            idata = pm.sample()
+
+        with uncensor(model):
+            idata_pp = pm.sample_posterior_predictive(idata, var_names=["y_uncensored"])
+
+        az.summary(idata_pp)
+    """
+    fg = fgraph_from_model(model)
+
+    (_, nodes_changed, *_) = uncensor_rewrite.apply(fg)
+    if not nodes_changed:
+        raise RuntimeError("No censored variables were replaced by uncensored counterparts")
+
+    return model_from_fgraph(fg)
+
+
+@node_rewriter(tracks=[ModelValuedVar])
+def forecast_timeseries_node_rewrite(fgraph: FunctionGraph, node):
+    """Rewrite that replaces timeseries variables by new ones starting at the last timepoint(s)."""
+
+    (
+        timeseries_rv,
+        value,
+        *dims,
+    ) = node.inputs
+    if not isinstance(timeseries_rv.owner.op, AutoRegressiveRV):
+        return
+
+    forecast_steps = get_var_by_name(fgraph.inputs, "forecast_steps_")
+    if len(forecast_steps) != 1:
+        return False
+
+    forecast_steps = forecast_steps[0]
+
+    op = timeseries_rv.owner.op
+    model_rv = node.outputs[0]
+
+    # We cannot reference the variable we are planning to replace
+    # Or it will introduce circularities in the graph
+    # FIXME: This special logic shouldn't be needed for ObservedRVs
+    # but PyMC does not allow one to not resample observed.
+    # We hack around by conditioning on the value variable directly,
+    # even though that should not be part of the generative graph...
```

Comment:
I disagree. This special logic IS needed for ObservedRVs because forecasting is different than posterior predictive.

Posterior predictive does something like this:

$$
y_{i} ~ P(Y|y_{1}^{obs}, ..., y_{n}^{obs})
$$

draw other potential values of the $y_{i}^{obs}$ observations from the probability distribution of the observed variable conditioned on the previous set of observations. Essentially, this is agnostic of temporal order and is intended for resampling of the entire set of observations.

Forecasting does this:

$$
y_{t+1} ~ P(Y_{t+1}|y_{1}^{obs}, ..., y_{n}^{obs})
$$

So it samples from a different conditional distribution than the posterior predictive. That's why we do need to manually add the past `y_{1}^{obs}, ..., y_{n}^{obs}` values explicitly.

Reply:
Observed variables are just FreeRVs that had constant values during sampling. If I can condition on FreeRVs traced values I should also be able to condition on ObservedRV traced values (constant) without having to hack across the value variable representation on the forward graph.

Similarly, if I had a Deterministic based on an ObservedRV (as happens with imputation), why can't I now make predictions based on those traced values?